File: app/src/main/java/com/parabola/newtone/ui/router/MainRouterImpl.java
package com.parabola.newtone.ui.router;

import android.app.SearchManager;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.core.content.FileProvider;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;

import com.parabola.domain.model.Track;
import com.parabola.newtone.BuildConfig;
import com.parabola.newtone.R;
import com.parabola.newtone.ui.activity.MainActivity;
import com.parabola.newtone.ui.dialog.ChoosePlaylistDialog;
import com.parabola.newtone.ui.dialog.CreatePlaylistDialog;
import com.parabola.newtone.ui.dialog.DeleteTrackDialog;
import com.parabola.newtone.ui.dialog.RenamePlaylistDialog;
import com.parabola.newtone.ui.dialog.SleepTimerDialog;
import com.parabola.newtone.ui.dialog.SortingDialog;
import com.parabola.newtone.ui.dialog.TimeToSleepInfoDialog;
import com.parabola.newtone.ui.dialog.TrackAdditionalInfoDialog;
import com.parabola.newtone.ui.dialog.fx.AudioEffectsDialog;
import com.parabola.newtone.ui.dialog.fx.EqPresetsSelectorDialog;
import com.parabola.newtone.ui.fragment.AlbumFragment;
import com.parabola.newtone.ui.fragment.ArtistFragment;
import com.parabola.newtone.ui.fragment.ArtistTracksFragment;
import com.parabola.newtone.ui.fragment.FolderFragment;
import com.parabola.newtone.ui.fragment.PlayerFragment;
import com.parabola.newtone.ui.fragment.SearchFragment;
import com.parabola.newtone.ui.fragment.playlist.FavoritesPlaylistFragment;
import com.parabola.newtone.ui.fragment.playlist.FoldersListFragment;
import com.parabola.newtone.ui.fragment.playlist.PlaylistFragment;
import com.parabola.newtone.ui.fragment.playlist.QueueFragment;
import com.parabola.newtone.ui.fragment.playlist.RecentlyAddedPlaylistFragment;
import com.parabola.newtone.ui.fragment.settings.AlbumItemDisplaySettingFragment;
import com.parabola.newtone.ui.fragment.settings.ArtistItemDisplaySettingFragment;
import com.parabola.newtone.ui.fragment.settings.ColorThemeSelectorFragment;
import com.parabola.newtone.ui.fragment.settings.ExcludedFoldersFragment;
import com.parabola.newtone.ui.fragment.settings.SettingFragment;
import com.parabola.newtone.ui.fragment.settings.TrackItemDisplaySettingFragment;
import com.parabola.newtone.ui.fragment.settings.dialog.IsaacNewtoneDialog;
import com.parabola.newtone.ui.fragment.start.StartFragment;
import com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelState;

import java.io.File;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;

public final class MainRouterImpl implements MainRouter {
    private MainActivity activity;
    private StartFragment firstFragment;

    @Override
    public void setActivity(MainActivity activity) {
        this.activity = activity;

        firstFragment = getInstance(StartFragment.class);
        if (firstFragment == null) {
            firstFragment = new StartFragment();

            activity.getSupportFragmentManager().beginTransaction()
                    .add(R.id.nav_host_fragment, firstFragment)
                    .commit();
        }
    }

    @Override
    public void clearActivity() {
        this.activity = null;
        this.firstFragment = null;
    }

    @Override
    public Fragment currentFragment() {
        Fragment currentFragment;
        if (isRoot()) {
            currentFragment = firstFragment.getCurrentSelectedFragment();
        } else {
            currentFragment = activity.getSupportFragmentManager()
                    .getPrimaryNavigationFragment();
        }
        return currentFragment;
    }

    @Override
    public void collapseBottomSlider() {
        activity.setBottomSliderPanelState(PanelState.COLLAPSED);
    }

    @Override
    public void goToTab(int tabNumber, boolean smoothScroll) {
        firstFragment.goToTab(tabNumber, smoothScroll);
    }

    @Override
    public void scrollOnTabTrackToCurrentTrack() {
        firstFragment.scrollOnTabTrackToCurrentTrack();
    }


    @Override
    public void goToArtistInTab(int artistId) {
        firstFragment.scrollToArtistInTab(artistId);
    }


    @Override
    public void goToAlbumInTab(int albumId) {
        firstFragment.scrollToAlbumInTab(albumId);
    }

    @Override
    public <T extends Fragment> boolean hasInstanceInStack(Class<T> fragment) {
        for (Fragment f : activity.getSupportFragmentManager().getFragments()) {
            if (fragment.isInstance(f)) {
                return true;
            }
        }

        return false;
    }

    @Nullable
    private <T extends Fragment> T getInstance(Class<T> fragment) {
        for (Fragment f : activity.getSupportFragmentManager().getFragments()) {
            if (fragment.isInstance(f)) {
                return fragment.cast(f);
            }
        }
        return null;
    }

    @Override
    public void backToRoot() {
        while (!isRoot()) {
            activity.getSupportFragmentManager().popBackStackImmediate();
        }
    }

    @Override
    public boolean isRoot() {
        return activity.getSupportFragmentManager().getBackStackEntryCount() == 0;
    }

    @Override
    public void goBack() {
        activity.getSupportFragmentManager()
                .popBackStackImmediate();
    }


    @Override
    public void openArtist(int artistId) {
        ArtistFragment artistFragment = ArtistFragment.newInstance(artistId);

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, artistFragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(artistFragment)
                .commit();
    }

    @Override
    public void openArtistFromBackStackIfAvailable(int artistId) {
        while (!isRoot()) {
            Fragment currentFragment = currentFragment();
            if (currentFragment instanceof ArtistFragment) {
                ArtistFragment artistFragment = (ArtistFragment) currentFragment;
                if (artistFragment.getArtistId() == artistId)
                    return;
            }

            goBack();
        }

        openArtist(artistId);
    }


    @Override
    public void openAlbum(int albumId) {
        AlbumFragment fragment = AlbumFragment.newInstance(albumId);

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openAlbumFromBackStackIfAvailable(int albumId) {
        while (!isRoot()) {
            Fragment currentFragment = currentFragment();
            if (currentFragment instanceof AlbumFragment) {
                AlbumFragment albumFragment = (AlbumFragment) currentFragment;
                if (albumFragment.getAlbumId() == albumId)
                    return;
            }

            goBack();
        }

        openAlbum(albumId);
    }


    @Override
    public void openPlaylist(int playlistId) {
        PlaylistFragment playlistFragment = PlaylistFragment.newInstance(playlistId);

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, playlistFragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(playlistFragment)
                .commit();
    }

    @Override
    public void openRequestStoragePermissionScreen() {
        activity.requestStoragePermission();
    }

    @Override
    public void openSearchScreen() {
        SearchFragment searchFragment = SearchFragment.newInstance();
        activity.getSupportFragmentManager().beginTransaction()
                .add(R.id.nav_host_fragment, searchFragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(searchFragment)
                .commit();
    }

    @Override
    public void openSettings() {
        SettingFragment fragment = new SettingFragment();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openSettingsIfAvailable() {
        while (!isRoot()) {
            Fragment currentFragment = currentFragment();
            if (currentFragment instanceof SettingFragment) {
                return;
            }

            goBack();
        }

        openSettings();
    }


    @Override
    public void openColorThemeSelectorSettings() {
        ColorThemeSelectorFragment fragment = new ColorThemeSelectorFragment();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openExcludedFolders() {
        FragmentTransaction fragmentTransaction = activity.getSupportFragmentManager().beginTransaction();
        //закрываются все окна, за исключением настроек, плеера и начального экрана
        //чтобы после исключения папок не было ненужных ошибок
        for (Fragment fragment : activity.getSupportFragmentManager().getFragments()) {
            if (!(fragment instanceof SettingFragment)
                    && !(fragment instanceof StartFragment)
                    && !(fragment instanceof PlayerFragment)) {
                fragmentTransaction.remove(fragment);
            }
        }
        fragmentTransaction.commit();


        ExcludedFoldersFragment fragment = ExcludedFoldersFragment.newInstance();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }


    @Override
    public void openTrackItemDisplaySettings() {
        TrackItemDisplaySettingFragment fragment = TrackItemDisplaySettingFragment.newInstance();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openAlbumItemDisplaySettings() {
        AlbumItemDisplaySettingFragment fragment = AlbumItemDisplaySettingFragment.newInstance();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openArtistItemDisplaySettings() {
        ArtistItemDisplaySettingFragment fragment = ArtistItemDisplaySettingFragment.newInstance();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openRecentlyAdded() {
        RecentlyAddedPlaylistFragment fragment = new RecentlyAddedPlaylistFragment();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openFavourites() {
        FavoritesPlaylistFragment fragment = new FavoritesPlaylistFragment();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openFavouritesFromBackStackIfAvailable() {
        while (!isRoot()) {
            Fragment currentFragment = currentFragment();
            if (currentFragment instanceof FavoritesPlaylistFragment) {
                return;
            }

            goBack();
        }

        openFavourites();
    }

    @Override
    public void openQueue() {
        QueueFragment fragment = new QueueFragment();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openQueueFromBackStackIfAvailable() {
        while (!isRoot()) {
            Fragment currentFragment = currentFragment();
            if (currentFragment instanceof QueueFragment) {
                return;
            }

            goBack();
        }

        openQueue();
    }

    @Override
    public void openFoldersList() {
        FoldersListFragment fragment = new FoldersListFragment();

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openFolder(String folderPath) {
        Bundle args = new Bundle();
        args.putString("folderPath", folderPath);

        FolderFragment fragment = new FolderFragment();
        fragment.setArguments(args);

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openArtistTracks(int artistId) {
        Bundle args = new Bundle();
        args.putInt("artistId", artistId);

        ArtistTracksFragment fragment = new ArtistTracksFragment();
        fragment.setArguments(args);

        activity.getSupportFragmentManager().beginTransaction()
                .setCustomAnimations(R.anim.anim_in, R.anim.anim_out, R.anim.anim_in, R.anim.anim_out)
                .add(R.id.nav_host_fragment, fragment)
                .addToBackStack(null)
                .setPrimaryNavigationFragment(fragment)
                .commit();
    }

    @Override
    public void openCreatePlaylistDialog() {
        CreatePlaylistDialog dialogFragment = CreatePlaylistDialog.newInstance();
        dialogFragment.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openRenamePlaylistDialog(int playlistId) {
        RenamePlaylistDialog dialogFragment = RenamePlaylistDialog.newInstance(playlistId);
        dialogFragment.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openStartSleepTimerDialog() {
        SleepTimerDialog dialogFragment = SleepTimerDialog.newInstance();
        dialogFragment.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openSleepTimerInfoDialog() {
        TimeToSleepInfoDialog dialogFragment = TimeToSleepInfoDialog.newInstance();
        dialogFragment.show(activity.getSupportFragmentManager(), null);
    }


    @Override
    public void openAddToPlaylistDialog(int... trackIds) {
        ChoosePlaylistDialog dialogFragment = ChoosePlaylistDialog.newInstance(trackIds);
        dialogFragment.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openSortingDialog(String sortingListType) {
        SortingDialog dialogFragment = SortingDialog.newInstance(sortingListType);
        dialogFragment.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openTrackAdditionInfo(int trackId) {
        TrackAdditionalInfoDialog dialogFragment = TrackAdditionalInfoDialog.newInstance(trackId);
        dialogFragment.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openAudioEffectsDialog() {
        AudioEffectsDialog dialog = new AudioEffectsDialog();
        dialog.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openEqPresetsSelectorDialog() {
        EqPresetsSelectorDialog dialog = new EqPresetsSelectorDialog();
        dialog.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openNewtoneDialog() {
        IsaacNewtoneDialog dialog = new IsaacNewtoneDialog();
        dialog.show(activity.getSupportFragmentManager(), null);
    }

    @Override
    public void openDeleteTrackDialog(int trackId) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            activity.showRequestForDeletingTrack(trackId);
        } else {
            DeleteTrackDialog dialog = DeleteTrackDialog.newInstance(trackId);
            dialog.show(activity.getSupportFragmentManager(), null);
        }
    }

    @Override
    public void openLyricsSearch(Track track) {
        String searchQuery = String.format("%s %s lyrics", track.getArtistName(), track.getTitle());
        Intent intent = new Intent(Intent.ACTION_WEB_SEARCH)
                .putExtra(SearchManager.QUERY, searchQuery);
        activity.startActivity(intent);
    }

    @Override
    public void openContactDevelopersViaEmail() {
        Intent sendMailIntent = new Intent(Intent.ACTION_SENDTO);
        sendMailIntent.setData(Uri.parse("mailto:")); // only email apps should handle this
        sendMailIntent.putExtra(Intent.EXTRA_EMAIL, new String[]{"zizik.zizik@gmail.com"});
        sendMailIntent.putExtra(Intent.EXTRA_SUBJECT, "Newtone. Android");

        String intentTitle = activity.getString(R.string.setting_contact_developers_intent_title);
        activity.startActivity(Intent.createChooser(sendMailIntent, intentTitle));
    }

    @Override
    public void openPrivacyPolicyWebPage() {
        String url = "https://github.com/parabola47/privacy_policy/blob/master/newtone_pp.md";
        Intent i = new Intent(Intent.ACTION_VIEW)
                .setData(Uri.parse(url));
        activity.startActivity(i);
    }

    @Override
    public void openShareTrack(String trackFilePath) {
        Uri uri = FileProvider.getUriForFile(activity.getApplication(), BuildConfig.APPLICATION_ID, new File(trackFilePath));
        Intent share = new Intent(Intent.ACTION_SEND)
                .putExtra(Intent.EXTRA_STREAM, uri)
                .setType("audio/*");
        activity.startActivity(Intent.createChooser(share, activity.getString(R.string.track_menu_share_track_intent_title)));
    }


    private final BehaviorSubject<Float> slidePanelOffsetUpdates = BehaviorSubject.createDefault(0f);
    private final BehaviorSubject<PanelState> slidePanelStateUpdates = BehaviorSubject.createDefault(PanelState.COLLAPSED);

    @Override
    public void setBottomSlidePanelOffset(float offset) {
        if (offset < 0f || offset > 1f) {
            return;
        }
        slidePanelOffsetUpdates.onNext(offset);
    }

    @Override
    public Observable<Float> observeSlidePanelOffset() {
        return slidePanelOffsetUpdates;
    }

    @Override
    public void setBottomSlidePanelState(PanelState state) {
        slidePanelStateUpdates.onNext(state);
    }

    @Override
    public Observable<PanelState> observeSlidePanelState() {
        return slidePanelStateUpdates;
    }

}

File: app/src/main/java/com/parabola/newtone/ui/router/MainRouterImpl.kt
package com.parabola.newtone.ui.router

import android.app.SearchManager
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.view.Gravity
import android.widget.Toast
import androidx.core.content.FileProvider
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.Fragment
import com.parabola.domain.model.Track
import com.parabola.newtone.BuildConfig
import com.parabola.newtone.R
import com.parabola.newtone.presentation.DeleteTrackDialog
import com.parabola.newtone.presentation.album.AlbumFragment
import com.parabola.newtone.presentation.artist.ArtistFragment
import com.parabola.newtone.presentation.artisttracks.ArtistTracksFragment
import com.parabola.newtone.presentation.playlist.chooseplaylist.ChoosePlaylistDialog
import com.parabola.newtone.presentation.playlist.createplaylist.CreatePlaylistDialog
import com.parabola.newtone.presentation.folder.FolderFragment
import com.parabola.newtone.presentation.playlist.favourites.FavouritesPlaylistFragment
import com.parabola.newtone.presentation.playlist.folderslist.FoldersListFragment
import com.parabola.newtone.presentation.playlist.playlist.PlaylistFragment
import com.parabola.newtone.presentation.playlist.queue.QueueFragment
import com.parabola.newtone.presentation.playlist.recentlyadded.RecentlyAddedPlaylistFragment
import com.parabola.newtone.presentation.mainactivity.MainActivity
import com.parabola.newtone.ui.dialog.*
import com.parabola.newtone.ui.dialog.fx.AudioEffectsDialog
import com.parabola.newtone.ui.dialog.fx.EqPresetsSelectorDialog
import com.parabola.newtone.presentation.settings.dialog.IsaacNewtoneDialog
import com.parabola.newtone.presentation.main.start.StartFragment
import com.parabola.newtone.presentation.player.PlayerFragment
import com.parabola.newtone.presentation.player.timetosleepinfo.TimeToSleepInfoDialog
import com.parabola.newtone.presentation.playlist.renameplaylist.RenamePlaylistDialog
import com.parabola.newtone.presentation.search.SearchFragment
import com.parabola.newtone.presentation.settings.AlbumItemDisplaySettingFragment
import com.parabola.newtone.presentation.settings.ArtistItemDisplaySettingFragment
import com.parabola.newtone.presentation.settings.ColorThemeSelectorFragment
import com.parabola.newtone.presentation.settings.ExcludedFoldersFragment
import com.parabola.newtone.presentation.settings.SettingFragment
import com.parabola.newtone.presentation.settings.TrackItemDisplaySettingFragment
import com.sothree.slidinguppanel.SlidingUpPanelLayout.PanelState
import io.reactivex.Observable
import io.reactivex.subjects.BehaviorSubject
import java.io.File
import java.util.*

class MainRouterImpl : MainRouter {
    private var activity: MainActivity? = null
    private var firstFragment: StartFragment? = null

    override fun setActivity(activity: MainActivity) {
        this.activity = activity

        firstFragment = getInstance(StartFragment::class.java)
        if (firstFragment == null) {
            firstFragment = StartFragment()
            activity.supportFragmentManager.beginTransaction()
                .add(R.id.nav_host_fragment, firstFragment!!)
                .commit()
        }
    }

    override fun clearActivity() {
        activity = null
        firstFragment = null
    }

    override fun showToast(toastText: String, longLength: Boolean, showAtCenter: Boolean) {
        val toast = Toast.makeText(
            activity,
            toastText,
            if (longLength) Toast.LENGTH_LONG else Toast.LENGTH_SHORT
        )
        if (showAtCenter) {
            toast.setGravity(Gravity.CENTER, 0, 0)
        }

        toast.show()
    }

    override fun currentFragment(): Fragment? {
        return if (isRoot) firstFragment?.currentSelectedFragment
        else activity?.supportFragmentManager?.primaryNavigationFragment
    }

    override fun collapseBottomSlider() {
        activity?.setBottomSliderPanelState(PanelState.COLLAPSED)
    }

    override fun goToTab(tabNumber: Int, smoothScroll: Boolean) {
        firstFragment?.goToTab(tabNumber, smoothScroll)
    }

    override fun scrollOnTabTrackToCurrentTrack() {
        firstFragment?.scrollOnTabTrackToCurrentTrack()
    }

    override fun goToArtistInTab(artistId: Int) {
        firstFragment?.scrollToArtistInTab(artistId)
    }

    override fun goToAlbumInTab(albumId: Int) {
        firstFragment?.scrollToAlbumInTab(albumId)
    }

    override fun <T : Fragment> hasInstanceInStack(fragment: Class<T>): Boolean {
        for (f in activity?.supportFragmentManager?.fragments ?: Collections.emptyList()) {
            if (fragment.isInstance(f)) {
                return true
            }
        }
        return false
    }

    private fun <T : Fragment> getInstance(fragment: Class<T>): T? {
        for (f in activity?.supportFragmentManager?.fragments ?: Collections.emptyList()) {
            if (fragment.isInstance(f)) {
                return fragment.cast(f)
            }
        }
        return null
    }

    override fun backToRoot() {
        while (!isRoot) {
            activity?.supportFragmentManager?.popBackStackImmediate()
        }
    }

    override val isRoot: Boolean
        get() = activity?.supportFragmentManager?.backStackEntryCount == 0

    override fun goBack() {
        activity?.supportFragmentManager?.popBackStackImmediate()
    }

    override fun openArtist(artistId: Int) {
        openFragment(ArtistFragment.newInstance(artistId))
    }

    override fun openArtistFromBackStackIfAvailable(artistId: Int) {
        while (!isRoot) {
            val currentFragment = currentFragment()
            if (currentFragment is ArtistFragment) {
                if (currentFragment.artistId == artistId) return
            }
            goBack()
        }
        openArtist(artistId)
    }

    override fun openAlbum(albumId: Int) {
        openFragment(AlbumFragment.newInstance(albumId))
    }

    override fun openAlbumFromBackStackIfAvailable(albumId: Int) {
        while (!isRoot) {
            val currentFragment = currentFragment()
            if (currentFragment is AlbumFragment) {
                if (currentFragment.albumId == albumId) return
            }
            goBack()
        }
        openAlbum(albumId)
    }

    override fun openPlaylist(playlistId: Int) {
        openFragment(PlaylistFragment.newInstance(playlistId))
    }

    override fun openRequestStoragePermissionScreen() {
        activity?.requestStoragePermission()
    }

    override fun openSearchScreen() {
        openFragment(SearchFragment.newInstance(), false)
    }

    override fun openSettings() {
        openFragment(SettingFragment())
    }

    override fun openSettingsIfAvailable() {
        while (!isRoot) {
            val currentFragment = currentFragment()
            if (currentFragment is SettingFragment) {
                return
            }
            goBack()
        }
        openSettings()
    }

    override fun openColorThemeSelectorSettings() {
        openFragment(ColorThemeSelectorFragment())
    }

    override fun openExcludedFolders() {
        val fragmentTransaction = activity!!.supportFragmentManager.beginTransaction()
        //закрываются все окна, за исключением настроек, плеера и начального экрана
        //чтобы после исключения папок не было ненужных ошибок
        for (fragment in activity?.supportFragmentManager?.fragments ?: Collections.emptyList()) {
            if (fragment !is SettingFragment
                && fragment !is StartFragment
                && fragment !is PlayerFragment
            ) {
                fragmentTransaction.remove(fragment)
            }
        }
        fragmentTransaction.commit()
        openFragment(ExcludedFoldersFragment.newInstance())
    }

    override fun openTrackItemDisplaySettings() {
        openFragment(TrackItemDisplaySettingFragment.newInstance())
    }

    override fun openAlbumItemDisplaySettings() {
        openFragment(AlbumItemDisplaySettingFragment.newInstance())
    }

    override fun openArtistItemDisplaySettings() {
        openFragment(ArtistItemDisplaySettingFragment.newInstance())
    }

    override fun openRecentlyAdded() {
        openFragment(RecentlyAddedPlaylistFragment())
    }

    override fun openFavourites() {
        openFragment(FavouritesPlaylistFragment())
    }

    override fun openFavouritesFromBackStackIfAvailable() {
        while (!isRoot) {
            val currentFragment = currentFragment()
            if (currentFragment is FavouritesPlaylistFragment) {
                return
            }
            goBack()
        }
        openFavourites()
    }

    override fun openQueue() {
        openFragment(QueueFragment())
    }

    override fun openQueueFromBackStackIfAvailable() {
        while (!isRoot) {
            val currentFragment = currentFragment()
            if (currentFragment is QueueFragment) {
                return
            }
            goBack()
        }
        openQueue()
    }

    override fun openFoldersList() {
        openFragment(FoldersListFragment())
    }

    override fun openFolder(folderPath: String) {
        openFragment(FolderFragment.newInstance(folderPath))
    }

    override fun openArtistTracks(artistId: Int) {
        openFragment(ArtistTracksFragment.newInstance(artistId))
    }

    override fun openCreatePlaylistDialog() {
        showDialogFragment(CreatePlaylistDialog.newInstance())
    }

    override fun openRenamePlaylistDialog(playlistId: Int) {
        showDialogFragment(RenamePlaylistDialog.newInstance(playlistId))
    }

    override fun openStartSleepTimerDialog() {
        showDialogFragment(SleepTimerDialog.newInstance())
    }

    override fun openSleepTimerInfoDialog() {
        showDialogFragment(TimeToSleepInfoDialog.newInstance())
    }

    override fun openAddToPlaylistDialog(vararg trackIds: Int) {
        showDialogFragment(ChoosePlaylistDialog.newInstance(*trackIds))
    }

    override fun openSortingDialog(sortingListType: String) {
        showDialogFragment(SortingDialog.newInstance(sortingListType))
    }

    override fun openTrackAdditionInfo(trackId: Int) {
        showDialogFragment(TrackAdditionalInfoDialog.newInstance(trackId))
    }

    override fun openAudioEffectsDialog() {
        showDialogFragment(AudioEffectsDialog())
    }

    override fun openEqPresetsSelectorDialog() {
        showDialogFragment(EqPresetsSelectorDialog())
    }

    override fun openNewtoneDialog() {
        showDialogFragment(IsaacNewtoneDialog())
    }

    override fun openDeleteTrackDialog(trackId: Int) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            activity!!.showRequestForDeletingTrack(trackId)
        } else {
            showDialogFragment(DeleteTrackDialog.newInstance(trackId))
        }
    }

    override fun openLyricsSearch(track: Track) {
        val searchQuery = String.format("%s %s lyrics", track.artistName, track.title)
        val intent = Intent(Intent.ACTION_WEB_SEARCH)
            .putExtra(SearchManager.QUERY, searchQuery)
        startActivity(intent)
    }

    override fun openContactDevelopersViaEmail() {
        val sendMailIntent = Intent(Intent.ACTION_SENDTO).apply {
            data = Uri.parse("mailto:") // only email apps should handle this
            putExtra(Intent.EXTRA_EMAIL, arrayOf("zizik.zizik@gmail.com"))
            putExtra(Intent.EXTRA_SUBJECT, "Newtone. Android")
        }
        Intent.createChooser(
            sendMailIntent,
            activity!!.getString(R.string.setting_contact_developers_intent_title)
        )

        activity?.apply {
            val intentTitle = getString(R.string.setting_contact_developers_intent_title)
            startActivity(Intent.createChooser(sendMailIntent, intentTitle))
        }
    }

    override fun openPrivacyPolicyWebPage() {
        val url = "https://github.com/parabola47/privacy_policy/blob/master/newtone_pp.md"
        val intent = Intent(Intent.ACTION_VIEW)
            .setData(Uri.parse(url))
        startActivity(intent)
    }

    override fun openShareTrack(filePath: String) {
        val uri = FileProvider.getUriForFile(
            activity!!.application,
            BuildConfig.APPLICATION_ID,
            File(filePath)
        )
        val share = Intent(Intent.ACTION_SEND)
            .putExtra(Intent.EXTRA_STREAM, uri)
            .setType("audio/*")
        val intent = Intent.createChooser(
            share,
            activity!!.getString(R.string.track_menu_share_track_intent_title)
        )
        startActivity(intent)
    }

    private fun startActivity(intent: Intent) {
        activity?.startActivity(intent)
    }

    private fun openFragment(fragment: Fragment, withAnimation: Boolean = true) {
        activity?.supportFragmentManager?.beginTransaction()?.apply {
            if (withAnimation) {
                setCustomAnimations(
                    R.anim.anim_in,
                    R.anim.anim_out,
                    R.anim.anim_in,
                    R.anim.anim_out,
                )
            }
            add(R.id.nav_host_fragment, fragment)
            addToBackStack(null)
            setPrimaryNavigationFragment(fragment)
            commit()
        }
    }

    private fun showDialogFragment(dialog: DialogFragment) {
        activity?.supportFragmentManager?.let { dialog.show(it, null) }
    }

    private val slidePanelOffsetUpdates = BehaviorSubject.createDefault(0f)
    private val slidePanelStateUpdates = BehaviorSubject.createDefault(PanelState.COLLAPSED)
    override fun setBottomSlidePanelOffset(offset: Float) {
        if (offset < 0f || offset > 1f) {
            return
        }
        slidePanelOffsetUpdates.onNext(offset)
    }

    override fun observeSlidePanelOffset(): Observable<Float> = slidePanelOffsetUpdates

    override fun setBottomSlidePanelState(state: PanelState) {
        slidePanelStateUpdates.onNext(state)
    }

    override fun observeSlidePanelState(): Observable<PanelState> = slidePanelStateUpdates
}
